File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/V086Controller.java
package org.emulinker.kaillera.controller.v086;

import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableMap;
import com.google.common.flogger.FluentLogger;
import java.net.InetSocketAddress;
import java.util.*;
import java.util.concurrent.*;
import javax.inject.Inject;
import javax.inject.Singleton;
import org.apache.commons.configuration.*;
import org.emulinker.config.RuntimeFlags;
import org.emulinker.kaillera.access.AccessManager;
import org.emulinker.kaillera.controller.KailleraServerController;
import org.emulinker.kaillera.controller.v086.action.*;
import org.emulinker.kaillera.controller.v086.protocol.*;
import org.emulinker.kaillera.model.*;
import org.emulinker.kaillera.model.event.*;
import org.emulinker.kaillera.model.exception.*;
import org.emulinker.net.*;

@Singleton
public final class V086Controller implements KailleraServerController {
  static final FluentLogger logger = FluentLogger.forEnclosingClass();

  static final int MAX_BUNDLE_SIZE = 9;

  private boolean isRunning = false;

  ThreadPoolExecutor threadPool;
  KailleraServer server;
  private String[] clientTypes;
  Map<Integer, V086ClientHandler> clientHandlers =
      new ConcurrentHashMap<Integer, V086ClientHandler>();

  private int portRangeStart;
  private int extraPorts;
  Queue<Integer> portRangeQueue = new ConcurrentLinkedQueue<Integer>();

  final ImmutableMap<Class<?>, V086ServerEventHandler> serverEventHandlers;
  final ImmutableMap<Class<?>, V086GameEventHandler> gameEventHandlers;
  final ImmutableMap<Class<?>, V086UserEventHandler> userEventHandlers;

  V086Action[] actions = new V086Action[25];

  private final V086ClientHandler.Factory v086ClientHandlerFactory;
  private final RuntimeFlags flags;

  @Inject
  V086Controller(
      KailleraServer server,
      ThreadPoolExecutor threadPool,
      AccessManager accessManager,
      Configuration config,
      LoginAction loginAction,
      ACKAction ackAction,
      ChatAction chatAction,
      CreateGameAction createGameAction,
      JoinGameAction joinGameAction,
      KeepAliveAction keepAliveAction,
      QuitGameAction quitGameAction,
      QuitAction quitAction,
      StartGameAction startGameAction,
      GameChatAction gameChatAction,
      GameKickAction gameKickAction,
      UserReadyAction userReadyAction,
      CachedGameDataAction cachedGameDataAction,
      GameDataAction gameDataAction,
      DropGameAction dropGameAction,
      CloseGameAction closeGameAction,
      GameStatusAction gameStatusAction,
      GameDesynchAction gameDesynchAction,
      PlayerDesynchAction playerDesynchAction,
      GameInfoAction gameInfoAction,
      GameTimeoutAction gameTimeoutAction,
      InfoMessageAction infoMessageAction,
      V086ClientHandler.Factory v086ClientHandlerFactory,
      RuntimeFlags flags) {
    this.v086ClientHandlerFactory = v086ClientHandlerFactory;
    this.flags = flags;
    this.threadPool = threadPool;
    this.server = server;
    this.clientTypes = config.getStringArray("controllers.v086.clientTypes.clientType");

    this.portRangeStart = config.getInt("controllers.v086.portRangeStart");
    this.extraPorts = config.getInt("controllers.v086.extraPorts", 0);
    int maxPort = 0;
    for (int i = portRangeStart; i <= (portRangeStart + server.getMaxUsers() + extraPorts); i++) {
      portRangeQueue.add(i);
      maxPort = i;
    }

    logger.atWarning().log(
        "Listening on UDP ports: "
            + portRangeStart
            + " to "
            + maxPort
            + ".  Make sure these ports are open in your firewall!");

    Preconditions.checkArgument(
        flags.v086BufferSize() > 0, "controllers.v086.bufferSize must be > 0");

    // array access should be faster than a hash and we won't have to create
    // a new Integer each time
    actions[UserInformation.ID] = loginAction;
    actions[ClientACK.ID] = ackAction;
    actions[Chat.ID] = chatAction;
    actions[CreateGame.ID] = createGameAction;
    actions[JoinGame.ID] = joinGameAction;
    actions[KeepAlive.ID] = keepAliveAction;
    actions[QuitGame.ID] = quitGameAction;
    actions[Quit.ID] = quitAction;
    actions[StartGame.ID] = startGameAction;
    actions[GameChat.ID] = gameChatAction;
    actions[GameKick.ID] = gameKickAction;
    actions[AllReady.ID] = userReadyAction;
    actions[CachedGameData.ID] = cachedGameDataAction;
    actions[GameData.ID] = gameDataAction;
    actions[PlayerDrop.ID] = dropGameAction;

    // setup the server event handlers
    serverEventHandlers =
        ImmutableMap.<Class<?>, V086ServerEventHandler>builder()
            .put(ChatEvent.class, chatAction)
            .put(GameCreatedEvent.class, createGameAction)
            .put(UserJoinedEvent.class, loginAction)
            .put(GameClosedEvent.class, closeGameAction)
            .put(UserQuitEvent.class, quitAction)
            .put(GameStatusChangedEvent.class, gameStatusAction)
            .build();

    // setup the game event handlers
    gameEventHandlers =
        ImmutableMap.<Class<?>, V086GameEventHandler>builder()
            .put(UserJoinedGameEvent.class, joinGameAction)
            .put(UserQuitGameEvent.class, quitGameAction)
            .put(GameStartedEvent.class, startGameAction)
            .put(GameChatEvent.class, gameChatAction)
            .put(AllReadyEvent.class, userReadyAction)
            .put(GameDataEvent.class, gameDataAction)
            .put(UserDroppedGameEvent.class, dropGameAction)
            .put(GameDesynchEvent.class, gameDesynchAction)
            .put(PlayerDesynchEvent.class, playerDesynchAction)
            .put(GameInfoEvent.class, gameInfoAction)
            .put(GameTimeoutEvent.class, gameTimeoutAction)
            .build();

    // setup the user event handlers
    userEventHandlers =
        ImmutableMap.<Class<?>, V086UserEventHandler>builder()
            .put(ConnectedEvent.class, ackAction)
            .put(InfoMessageEvent.class, infoMessageAction)
            .build();
  }

  @Override
  public String getVersion() {
    return "v086";
  }

  @Override
  public String[] getClientTypes() {
    return clientTypes;
  }

  @Override
  public KailleraServer getServer() {
    return server;
  }

  @Override
  public int getNumClients() {
    return clientHandlers.size();
  }

  @Override
  public int getBufferSize() {
    return flags.v086BufferSize();
  }

  public final ImmutableMap<Class<?>, V086ServerEventHandler> getServerEventHandlers() {
    return serverEventHandlers;
  }

  public final ImmutableMap<Class<?>, V086GameEventHandler> getGameEventHandlers() {
    return gameEventHandlers;
  }

  public final ImmutableMap<Class<?>, V086UserEventHandler> getUserEventHandlers() {
    return userEventHandlers;
  }

  public final V086Action[] getActions() {
    return actions;
  }

  public final Map<Integer, V086ClientHandler> getClientHandlers() {
    return clientHandlers;
  }

  public boolean isRunning() {
    return isRunning;
  }

  @Override
  public String toString() {
    return "V086Controller[clients=" + clientHandlers.size() + " isRunning=" + isRunning + "]";
  }

  @Override
  public int newConnection(InetSocketAddress clientSocketAddress, String protocol)
      throws ServerFullException, NewConnectionException {
    if (!isRunning) throw new NewConnectionException("Controller is not running");

    V086ClientHandler clientHandler = v086ClientHandlerFactory.create(clientSocketAddress, this);
    KailleraUser user = server.newConnection(clientSocketAddress, protocol, clientHandler);

    int boundPort = -1;
    int bindAttempts = 0;
    while (bindAttempts++ < 5) {
      Integer portInteger = portRangeQueue.poll();
      if (portInteger == null) {
        logger.atSevere().log("No ports are available to bind for: " + user);
      } else {
        int port = portInteger.intValue();
        logger.atInfo().log("Private port " + port + " allocated to: " + user);

        try {
          clientHandler.bind(port);
          boundPort = port;
          break;
        } catch (BindException e) {
          logger.atSevere().withCause(e).log("Failed to bind to port " + port + " for: " + user);
          logger.atFine().log(
              toString()
                  + " returning port "
                  + port
                  + " to available port queue: "
                  + (portRangeQueue.size() + 1)
                  + " available");
          portRangeQueue.add(port);
        }
      }

      try {
        // pause very briefly to give the OS a chance to free a port
        Thread.sleep(5);
      } catch (InterruptedException e) {
      }
    }

    if (boundPort < 0) {
      clientHandler.stop();
      throw new NewConnectionException("Failed to bind!");
    }

    clientHandler.start(user);
    return boundPort;
  }

  @Override
  public synchronized void start() {
    isRunning = true;
  }

  @Override
  public synchronized void stop() {
    isRunning = false;

    for (V086ClientHandler clientHandler : clientHandlers.values()) clientHandler.stop();

    clientHandlers.clear();
  }
}

File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/V086Controller.kt
package org.emulinker.kaillera.controller.v086

import com.google.common.flogger.FluentLogger
import java.net.InetSocketAddress
import java.net.SocketException
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentLinkedQueue
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.reflect.KClass
import kotlin.time.Duration.Companion.milliseconds
import kotlinx.coroutines.*
import org.apache.commons.configuration.Configuration
import org.emulinker.config.RuntimeFlags
import org.emulinker.kaillera.controller.KailleraServerController
import org.emulinker.kaillera.controller.v086.action.*
import org.emulinker.kaillera.controller.v086.protocol.*
import org.emulinker.kaillera.model.KailleraServer
import org.emulinker.kaillera.model.event.*
import org.emulinker.kaillera.model.exception.NewConnectionException
import org.emulinker.kaillera.model.exception.ServerFullException
import org.emulinker.net.UdpSocketProvider

private val logger = FluentLogger.forEnclosingClass()

/** High level logic for handling messages on a port. Not tied to an individual user. */
@Singleton
class V086Controller
    @Inject
    internal constructor(
        override var server: KailleraServer,
        config: Configuration,
        loginAction: LoginAction,
        ackAction: ACKAction,
        chatAction: ChatAction,
        createGameAction: CreateGameAction,
        joinGameAction: JoinGameAction,
        keepAliveAction: KeepAliveAction,
        quitGameAction: QuitGameAction,
        quitAction: QuitAction,
        startGameAction: StartGameAction,
        gameChatAction: GameChatAction,
        gameKickAction: GameKickAction,
        userReadyAction: UserReadyAction,
        cachedGameDataAction: CachedGameDataAction,
        gameDataAction: GameDataAction,
        dropGameAction: DropGameAction,
        closeGameAction: CloseGameAction,
        gameStatusAction: GameStatusAction,
        gameDesynchAction: GameDesynchAction,
        playerDesynchAction: PlayerDesynchAction,
        gameInfoAction: GameInfoAction,
        gameTimeoutAction: GameTimeoutAction,
        infoMessageAction: InfoMessageAction,
        private val v086ClientHandlerFactory: V086ClientHandler.Factory,
        flags: RuntimeFlags
    ) : KailleraServerController {
  var isRunning = false
    private set

  override val clientTypes: Array<String> =
      config.getStringArray("controllers.v086.clientTypes.clientType")

  override val clientHandlers: MutableMap<Int, V086ClientHandler> = ConcurrentHashMap()

  private val portRangeStart: Int = config.getInt("controllers.v086.portRangeStart")
  private val extraPorts: Int = config.getInt("controllers.v086.extraPorts", 0)

  var portRangeQueue: Queue<Int> = ConcurrentLinkedQueue()

  val serverEventHandlers: Map<KClass<out ServerEvent>, V086ServerEventHandler<Nothing>> =
      mapOf(
          ChatEvent::class to chatAction,
          GameCreatedEvent::class to createGameAction,
          UserJoinedEvent::class to loginAction,
          GameClosedEvent::class to closeGameAction,
          UserQuitEvent::class to quitAction,
          GameStatusChangedEvent::class to gameStatusAction,
      )
  val gameEventHandlers: Map<KClass<out GameEvent>, V086GameEventHandler<Nothing>> =
      mapOf(
          UserJoinedGameEvent::class to joinGameAction,
          UserQuitGameEvent::class to quitGameAction,
          GameStartedEvent::class to startGameAction,
          GameChatEvent::class to gameChatAction,
          AllReadyEvent::class to userReadyAction,
          GameDataEvent::class to gameDataAction,
          UserDroppedGameEvent::class to dropGameAction,
          GameDesynchEvent::class to gameDesynchAction,
          PlayerDesynchEvent::class to playerDesynchAction,
          GameInfoEvent::class to gameInfoAction,
          GameTimeoutEvent::class to gameTimeoutAction,
      )
  val userEventHandlers: Map<KClass<out UserEvent>, V086UserEventHandler<Nothing>> =
      mapOf(
          ConnectedEvent::class to ackAction,
          InfoMessageEvent::class to infoMessageAction,
      )

  var actions: Array<V086Action<*>?> = arrayOfNulls(25)

  override val version = "v086"

  override val numClients = clientHandlers.size

  override val bufferSize = flags.v086BufferSize

  override fun toString(): String {
    return "V086Controller[clients=${clientHandlers.size} isRunning=$isRunning]"
  }

  /**
   * Receives new connections and delegates to a new V086ClientHandler instance for communication
   * over a separate port.
   */
  @OptIn(DelicateCoroutinesApi::class) // For GlobalScope.
  @Throws(ServerFullException::class, NewConnectionException::class)
  override suspend fun newConnection(
      udpSocketProvider: UdpSocketProvider, clientSocketAddress: InetSocketAddress, protocol: String
  ): Int {
    if (!isRunning) throw NewConnectionException("Controller is not running")
    logger
        .atFine()
        .log("Creating new connection for address %d, protocol %s", clientSocketAddress, protocol)

    val clientHandler = v086ClientHandlerFactory.create(clientSocketAddress, this)
    val user = server.newConnection(clientSocketAddress, protocol, clientHandler)
    var boundPort: Int? = null
    var bindAttempts = 0
    while (bindAttempts++ < 5) {
      val portInteger = portRangeQueue.poll()
      if (portInteger == null) {
        logger.atSevere().log("No ports are available to bind for: $user")
      } else {
        val port = portInteger.toInt()
        logger.atInfo().log("Private port $port allocated to: $user")
        try {
          clientHandler.bind(udpSocketProvider, port)
          GlobalScope.launch(Dispatchers.IO) { clientHandler.run(coroutineContext) }
          boundPort = port
          break
        } catch (e: SocketException) {
          logger.atSevere().withCause(e).log("Failed to bind to port $port for: $user")
          logger
              .atFine()
              .log(
                  "$this returning port $port to available port queue: ${portRangeQueue.size + 1} available")
          portRangeQueue.add(port)
        }
      }
      // pause very briefly to give the OS a chance to free a port
      delay(5.milliseconds)
    }
    if (boundPort == null) {
      clientHandler.stop()
      throw NewConnectionException("Failed to bind!")
    }
    clientHandler.start(user)
    return boundPort
  }

  @Synchronized
  override fun start() {
    isRunning = true
  }

  @Synchronized
  override suspend fun stop() {
    isRunning = false
    clientHandlers.values.forEach { it.stop() }
    clientHandlers.clear()
  }

  init {
    var maxPort = 0
    for (i in portRangeStart..portRangeStart + server.maxUsers + extraPorts) {
      portRangeQueue.add(i)
      maxPort = i
    }
    logger
        .atWarning()
        .log(
            "Listening on UDP ports: $portRangeStart to $maxPort.  Make sure these ports are open in your firewall!")

    // array access should be faster than a hash and we won't have to create
    // a new Integer each time
    actions[UserInformation.ID.toInt()] = loginAction
    actions[ClientACK.ID.toInt()] = ackAction
    actions[Chat.ID.toInt()] = chatAction
    actions[CreateGame.ID.toInt()] = createGameAction
    actions[JoinGame.ID.toInt()] = joinGameAction
    actions[KeepAlive.ID.toInt()] = keepAliveAction
    actions[QuitGame.ID.toInt()] = quitGameAction
    actions[Quit.ID.toInt()] = quitAction
    actions[StartGame.ID.toInt()] = startGameAction
    actions[GameChat.ID.toInt()] = gameChatAction
    actions[GameKick.ID.toInt()] = gameKickAction
    actions[AllReady.ID.toInt()] = userReadyAction
    actions[CachedGameData.ID.toInt()] = cachedGameDataAction
    actions[GameData.ID.toInt()] = gameDataAction
    actions[PlayerDrop.ID.toInt()] = dropGameAction
  }

  companion object {
    const val MAX_BUNDLE_SIZE = 9
  }
}
